src/overview.md

```diff
@@ -1,3 +1,400 @@
+  [`librustc_parse`] to prepare for the next stage of the compile process. The
+  [`StringReader`] struct is used at this stage to perform a set of validations
+  and turn strings into interned symbols (_interning_ is discussed later).
+- The lexer has a small interface and doesn't depend directly on the
+  diagnostic infrastructure in `rustc`. Instead it provides diagnostics as plain
+  data which are emitted in `librustc_parse::lexer::mod` as real diagnostics.
+- The lexer preseves full fidelity information for both IDEs and proc macros.
+- The parser [translates the token stream from the lexer into an Abstract Syntax
+  Tree (AST)][parser].  It uses a recursive descent (top-down) approach to syntax
+  analysis. The crate entry points for the parser are the `Parser.parse_crate_mod()` and
+  `Parser.parse_mod()` methods found in `librustc_parse::parser::item`. The external
+  module parsing entry point is `librustc_expand::module::parse_external_mod`. And
+  the macro parser entry point is `rustc_expand::mbe::macro_parser::parse_nt`.
+- Parsing is performed with a set of `Parser` utility methods including `fn bump`,
+  `fn check`, `fn eat`, `fn expect`, `fn look_ahead`.
+- Parsing is organized by the semantic construct that is being parsed. Separate
+  `parse_*` methods can be found in `librustc_parse` `parser` directory. The source
+  file name follows the construct name. For example, the following files are found
+  in the parser:
+    - `expr.rs`
+    - `pat.rs`
+    - `ty.rs`
+    - `stmt.rs`
+- This naming scheme is used across many compiler stages. You will find
+  either a file or directory with the same name across the parsing, lowering,
+  type checking, HAIR lowering, and MIR building sources.
+- Macro expansion, AST validation, name resolution, and early linting takes place
+  during this stage of the compile process.
+- The parser uses the standard `DiagnosticBuilder` API for error handling, but we
+  try to recover, parsing a superset of Rust's grammar, while also emitting an error.
+- `rustc_ast::ast::{Crate, Mod, Expr, Pat, ...}` AST nodes are returned from the parser.
+- We then take the AST and [convert it to High-Level Intermediate
+  Representation (HIR)][hir]. This is a compiler-friendly representation of the
+  AST.  This involves a lot of desugaring of things like loops and `async fn`.
+- We use the HIR to do [type inference]. This is the process of automatic
+  detection of the type of an expression.
+- **TODO: Maybe some other things are done here? I think initial type checking
+  happens here? And trait solving?**
+- The HIR is then [lowered to Mid-Level Intermediate Representation (MIR)][mir].
+  - Along the way, we construct the HAIR, which is an even more desugared HIR.
+    HAIR is used for pattern and exhaustiveness checking. It is also more
+    convenient to convert into MIR than HIR is.
+- The MIR is used for [borrow checking].
+- **TODO: const eval fits in somewhere here I think**
```

Comment:
Should be confirmed and removed before merge
```suggestion

```

Reply:
added to the followup issue